influxdb3_client/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+use std::string::FromUtf8Error;
+
+use bytes::Bytes;
+use reqwest::{Body, IntoUrl, StatusCode};
+use secrecy::{ExposeSecret, Secret};
+use serde::Serialize;
+use url::Url;
+
+/// Primary error type for the [`Client`]
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("base URL error: {0}")]
+    BaseUrl(#[source] reqwest::Error),
+
+    #[error("request URL error: {0}")]
+    RequestUrl(#[from] url::ParseError),
+
+    #[error("failed to send /api/v3/write_lp request: {0}")]
+    WriteLpSend(#[source] reqwest::Error),
+
+    #[error("failed to read the API response bytes: {0}")]
+    Bytes(#[source] reqwest::Error),
+
+    #[error("failed to send /api/v3/query_sql request: {0}")]
+    QuerySqlSend(#[source] reqwest::Error),
+
+    #[error("invalid UTF8 in response: {0}")]
+    InvalidUtf8(#[from] FromUtf8Error),
+
+    #[error("server responded with error [{code}]: {message}")]
+    ApiError { code: StatusCode, message: String },
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// The InfluxDB 3.0 Client
+///
+/// For programmatic access to the HTTP API of InfluxDB 3.0
+#[derive(Debug, Clone)]
+pub struct Client {
+    /// The base URL for making requests to a running InfluxDB 3.0 server
+    base_url: Url,
+    /// The `Bearer` token to use for authenticating on each request to the server
+    auth_header: Option<Secret<String>>,
+    /// A [`reqwest::Client`] for handling HTTP requests
+    http_client: reqwest::Client,
+}
+
+impl Client {
+    /// Create a new [`Client`]
+    ///
+    /// A [`reqwest::Client`] is passed in order to allow for re-use of an existing
+    /// `reqwest::Client`.
+    pub fn new<U: IntoUrl>(base_url: U, http_client: reqwest::Client) -> Result<Self> {
+        Ok(Self {
+            base_url: base_url.into_url().map_err(Error::BaseUrl)?,
+            auth_header: None,
+            http_client,
+        })
+    }
+
+    /// Set the `Bearer` token that will be sent with each request to the server
+    pub fn with_auth_header<S: Into<String>>(mut self, auth_header: S) -> Self {
```

Comment:
I think the naming on this method is a bit off. I think it should be `with_auth_token`, or `with_bearer_token`.